nbclient/client.py

```diff
@@ -329,18 +329,14 @@ def start_kernel_manager(self):
         return self.km
 
     async def _async_cleanup_kernel(self):
+        now = self.shutdown_kernel == "immediate"
         try:
-            # Send a polite shutdown request
-            await ensure_async(self.kc.shutdown())
-            try:
-                # Queue the manager to kill the process, sometimes the built-in and above
-                # shutdowns have not been successful or called yet, so give a direct kill
-                # call here and recover gracefully if it's already dead.
-                await ensure_async(self.km.shutdown_kernel(now=True))
-            except RuntimeError as e:
-                # The error isn't specialized, so we have to check the message
-                if 'No kernel is running!' not in str(e):
-                    raise
+            # Queue the manager to kill the process, and recover gracefully if it's already dead.
+            await ensure_async(self.km.shutdown_kernel(now=now))
```

Comment:
I _think_ this will be ok. I would make sure you test the parallel execution and ensure that A) the graceful shutdown is degrading to ungraceful if it take too long and B) the kernels are shutting down without zombie processes left over when graceful exit has to revert to ungraceful. There used to be a lot of issues with graceful kernel shutdown in the headless executions as they weren't always cleaning up before the parent process exited. e.g. with a papermil CLI call

Reply:
If you can do some quick testing with parallel processing and make a situation where the kernel won't exit cleanly (e.g. the kernel is busy executing something else) that'd be ideal before merging.

Reply:
I've verified the graceful shutdown defaults to ungraceful by using a notebook with an atexit that sleeps, and also a thread that sleeps and doesn't join. Both cause the ungraceful SIGKILL. In neither case were there zombie processes; that said, if the notebook itself was starting up multiple processes and making them not clean up right, you could end up with zombies, but that's always been the case.

If a papermill CLI call or just direct nbclient call is made, and gets interrupted, I don't believe the exit is graceful. I thought the context manager would handle this, but it seems not. I'll look into why that is.

Not sure what "parallel execution" means. Could you please explain what you meant there?

Reply:
TIL asyncio and signal handling don't go together

Reply:
Thanks for doing some testing here. By parallel execution I mean something like:

```
import multiprocessing

def worker():
    # or just nbclient directly here tbh
    return pm.execute_notebook(input_path, output_path)

if __name__ == '__main__':
    jobs = []
    for i in range(5):
        p = multiprocessing.Process(target=worker)
        jobs.append(p)
        p.start()
    for job in jobs:
      job.join()
```

where there's a subprocess boundary and some parallelism. There's been lots of issues here around cleanup and termination so I wanted to double check it's alright.

Reply:
In the example, what would you worry about going wrong? If the notebook client isn't being shared, there _shouldn't_ be any issue. If all you're asking is that I test that out, no problem, but I can't really see there being issues there if nothing gets interrupted or killed or whatever.

IMO, nbclient should either guarantee an attempt is made to shutdown kernels when INTed or TERMed, which is what I'm looking into currently. The mashup of async and non-async code is making this pretty tricky, and one problem is all the cleanup code requires the ioloop which is stopped when either of those signals are received

Reply:
:facepalm: I've been mucking around with signals, but really atexit should do the job. But then I guess it would need to be in the main thread/process; also, iirc atexit doesn't play well with `multiprocessing`

Reply:
You're hitting all the reasons I didn't leave it graceful, unfortunately. `atexit` is best effort and does not run for subprocess calls. The testing parallel subworkflows is really about testing process exit patterns against the os._exit call it makes (which skips atexit). In python < 3.6 there were some rough edges that I remember being problematic under certain conditions as well. https://github.com/kuralabs/multiexit was an attempt to fix this but seems incomplete / only usable if you're guaranteed to be the entry process.

For async cleanup, maybe take a look at https://mail.python.org/pipermail/async-sig/2017-August/000374.html and see if that can be utilized here. Might simplify the problem substantially if so. Also it's ok if there's some extreme case where we can't cleanup -- you can't protect against some like OOM or SIGKILL -- but making an effort to reduce that to the fewest cases possible is worth it.

Reply:
Yeah atexit is not a solution in case of multiple processes. 

I had actually tried what's in that email for a while before giving up temporarily. What was happening was that the signal was somehow not being propagated after the signal handler was called, and it ended up with some weird exit behaviour like cleanup running while nbclient still executed the notebook, causing the execution to fail because the channels were closed and something timed out. I would have expected the signal to cause that to exit. 

I think the ideal way to handle this would be for the context manager to behave as usual and have the finally block be called when the program is interrupted or terminated. The problem there is not clear, but it appears to be a combination of async working strangely with KeyboardInterrupt, and the run_sync decorator